HSEARCH-1182 Add `excludePaths` to IndexedEmbedded (#3525)
hibernate/hibernate-search: documentation/src/main/asciidoc/reference/mapping-indexedembedded.asciidoc

or <<mapping-indexedembedded-prefix,prefix>>.
+
This takes precedence over `includeDepth` (see below).
+
Cannot be used in combination with `excludePaths` in the same `@IndexedEmbedded`.
`excludePaths`::
The paths of index fields from the indexed-embedded element that must *not* be embedded.
+
Provided paths must be relative to the indexed-embedded element,
i.e. they must not include its <<mapping-indexedembedded-name,name>>
or <<mapping-indexedembedded-prefix,prefix>>.
+
This takes precedence over `includeDepth` (see below).
+
Cannot be used in combination with `includePaths` in the same `@IndexedEmbedded`.
`includeDepth`::
The number of levels of indexed-embedded that will have all their fields included by default.
+
`includeDepth` is the number of `@IndexedEmbedded` that will be traversed
and for which all fields of the indexed-embedded element will be included,
even if these fields are not included explicitly through `includePaths`,

> unless a field is excluded explicitly through `excludePaths`:

Reviewer: ```suggestion
unless these fields are excluded explicitly through `excludePaths`:
```

even if these fields are not included explicitly through `includePaths`,
unless a field is excluded explicitly through `excludePaths`:
+
 * `includeDepth=0` means that fields of the indexed-embedded element are *not* included,
nor is any field of nested indexed-embedded elements,
unless these fields are included explicitly through `includePaths`.
 * `includeDepth=1` means that fields of the indexed-embedded element *are* included,
unless a field is excluded explicitly through `excludePaths`,
but *not* fields of nested indexed-embedded elements,
unless these fields are included explicitly through `includePaths`.
 * And so on.
+
The default value depends on the value of `includePaths`/`excludePaths` attributes:
+
* if both `includePaths` and `excludePaths` are empty, the default is `Integer.MAX_VALUE` (include all fields at every level)
* if `includePaths` is *not* empty and `excludePaths` is empty, the default is `0` (only include fields included explicitly)
* if `includePaths` is empty and `excludePaths` is *not* empty, the default is `Integer.MAX_VALUE` (include all fields except those that are excluded explicitly)

> * having both `includePaths` and `excludePaths` *not* empty is *not* allowed and will result in an exception.

Author: mmm 
> * if `includePaths` is empty and `excludePaths` is *not* empty, the default is `Integer.MAX_VALUE` (include all fields except those that are excluded explicitly)

if we have some exclude paths -- depth is max int... now I wonder if others would also miss it and if we should rephrase this 🙈

> * having both `includePaths` and `excludePaths` *not* empty is *not* allowed and will result in an exception.

Reviewer: But you also have:

> * if both `includePaths` and `excludePaths` are empty, the default is `Integer.MAX_VALUE` (include all fields at every level)

So as soon as `includePaths` is empty, the default is `Integer.MAX_VALUE`, whether `excludePaths` is empty or not. `excludePaths` does not matter.

> * having both `includePaths` and `excludePaths` *not* empty is *not* allowed and will result in an exception.

Author: well now 😄 if you put it like that...  it makes `excludePaths` feel... excluded 😆🤣

in order for that field to be included.
====

[NOTE]
.Mixing `includePaths` and `excludePaths` at different nesting levels
====
In general, it is possible to use `includePaths` and `excludePaths` at different levels of nested `@IndexedEmbedded`.
When doing so, keep in mind that the filter at each level can only reference reachable paths,

> i.e. filter cannot reference a path that, in general is there but, was made unreachable by a filter at a lower level that
> excluded at least one of this path's nodes (implicitly or explicitly).
> ====

Reviewer: ```suggestion
i.e. a filter cannot reference a path that was excluded by a nested `@IndexedEmbedded` (implicitly or explicitly).
====
```

in order for that field to be included.
====

[NOTE]
.Mixing `includePaths` and `excludePaths` at different nesting levels
====
In general, it is possible to use `includePaths` and `excludePaths` at different levels of nested `@IndexedEmbedded`.
When doing so, keep in mind that the resulting filters cannot be conflicting, i.e. including a property at one level

> and excluding it at the other will result in an exception being thrown.

Author: Ouch 😆  but that was the whole point of adding that `findConflictingIncludeExcludePaths` stuff ... 
What I was trying to address there was something like:
```java
class IndexedEntity {
	@IndexedEmbedded(excludePaths = "subIncluded.subIncludedString")
	Included included;
}

class Included {
	@KeywordField
	String includedString;
	// but we've excluded it explicitly already at the very top with exclude filter
	// should we include? exclude? no idea let's fail instead :)
	@IndexedEmbedded(includePaths = "subIncludedString") 
	SubIncluded subIncluded;
}

class SubIncluded {
	@KeywordField
	String subIncludedString;
}
```
and what I was thinking that we cannot simply leave it to path tracking or what we had before, since for example we could have something included/excluded because of a depth, but if it is included by depth and excluded by a excludePaths entry - it's ok we should allow it...

> and excluding it at the other will result in an exception being thrown.

Reviewer: Discussed here: https://hibernate.zulipchat.com/#narrow/stream/132092-hibernate-search-dev/topic/HSEARCH-1182.20IndexEmbedded.20exclude.20paths

----
include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/indexedembedded/onelevel/IndexedEmbeddedOneLevelIT.java[tags=programmatic]
----

> ====

Author: yeah, I think it comes from IDE (Settings -> Editor -> General -> Ensure line feed at file end on Save) and I might have it in git as well (an old habit I guess 🙈🥲)
I may or may not be wrong, but I think that the Eclipse formatter does this too, (or maybe I've added that in the config for it) and my idea was to have these blank lines added while we are modifying the files anyway so that when we do the mass formatting there'll be fewer changes...

> ====

Reviewer: > my idea was to have these blank lines added while we are modifying the files anyway so that when we do the mass formatting there'll be fewer changes...

Good point
